internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
+// performed for jobs being changed to a `completed` state, which we expect to
+// the vast common case under normal operation. The completer embeds an
+// AsyncCompleter to perform other non-`completed` state completions.
+type BatchCompleter struct {
+	baseservice.BaseService
+	startstop.BaseStartStop
+	withSubscribe
+
+	asyncCompleter   *AsyncCompleter // used for non-complete completions
+	debounceChan     *chanutil.DebouncedChan
+	exec             PartialExecutor
+	ready            chan struct{}
+	setStateParams   map[int64]*batchCompleterSetState
+	setStateParamsMu sync.Mutex
+}
+
+func NewBatchCompleter(archetype *baseservice.Archetype, exec PartialExecutor) *BatchCompleter {
+	return baseservice.Init(archetype, &BatchCompleter{
+		asyncCompleter: NewAsyncCompleter(archetype, exec),
+		exec:           exec,
+		setStateParams: make(map[int64]*batchCompleterSetState),
+	})
+}
+
+func (c *BatchCompleter) Start(ctx context.Context) error {
+	stopCtx, shouldStart, stopped := c.StartInit(ctx)
+	if !shouldStart {
+		return nil
+	}
+
+	c.ready = make(chan struct{})
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		c.Logger.InfoContext(ctx, c.Name+": Run loop started")
+		defer c.Logger.InfoContext(ctx, c.Name+": Run loop stopped")
 
-		func() {
-			c.subscribeFuncMu.Lock()
-			defer c.subscribeFuncMu.Unlock()
+		c.debounceChan = chanutil.NewDebouncedChan(stopCtx, 100*time.Millisecond)
 
-			if c.subscribeFunc != nil {
-				c.subscribeFunc(CompleterJobUpdated{Job: job, JobStats: stats})
+		close(c.ready)
+
+		for {
+			select {
+			case <-stopCtx.Done():
+				// Try to insert last batch before leaving. Note we use the
+				// original context so operations aren't immediately cancelled.
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
+
+				return
+			case <-c.debounceChan.C():
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
 			}
-		}()
+		}
+	}()
+
+	return nil
+}
 
+func (c *BatchCompleter) handleBatch(ctx context.Context) error {
+	var setStateBatch map[int64]*batchCompleterSetState
+	func() {
+		c.setStateParamsMu.Lock()
+		defer c.setStateParamsMu.Unlock()
+
+		setStateBatch = c.setStateParams
+
+		// Don't bother resetting the map if there's nothing to process,
+		// allowing the completer to idle efficiently.
+		if len(setStateBatch) > 0 {
+			c.setStateParams = make(map[int64]*batchCompleterSetState)
+		} else {
+			// Set nil to avoid a data race below in case the map is set as a
+			// new job comes in.
+			setStateBatch = nil
+		}
+	}()
+
+	if len(setStateBatch) < 1 {
 		return nil
-	})
+	}
+
+	// Insert a sub-batch with retries. Also helps reduce visual noise and
+	// increase readability of loop below.
+	insertSubBatch := func(batchID []int64, batchFinalizedAt []time.Time) ([]*rivertype.JobRow, error) {
```

Comment:
Slightly misnamed IMO because it's not _inserting_ anything. Maybe submit, update, or handle sub batch?

Reply:
Renamed to `completeSubBatch`.